implement corruption correcting recv

alek-p opened the pull request:
This patch implements a new type of zfs receive: corrective receive (-c). This type of recv is used to heal corrupted data when a replica of the data already exists (in the form of a sendfile for example).
Metadata can not be healed using a corrective receive.

This patch enables us to receive a send stream into an existing snapshot for the purpose of correcting data corruption. 

This is the updated version of the patch in https://github.com/zfsonlinux/zfs/pull/9323

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
In the past in the rare cases where ZFS has experienced permanent data corruption, full recovery of the dataset(s) has not always been possible even if replicas existed.
This patch makes recovery from permanent data corruption possible.

### Description
<!--- Describe your changes in detail -->
For every write record in the send stream, we read the corresponding block from disk and if that read fails with a checksum error we overwrite that block with data from the send stream.
After the data is healed we reread the block to make sure it's healed and remove the healed blocks form the corruption lists seen in zpool status.

To makes sure will have correctly matched the data in the send stream to the right dataset to heal there is a restriction that the GUID for the snapshot being received into must match the GUID in the send stream. There are likely several snapshots referring to the same potentially corrupted data so there may be many snapshots with the above condition holding that are able to heal a single block.

The other thing to point out is that we can only correct data. Specifically, we are only able to heal records of type DRR_WRITE.

To help with the review you can see my OpenZFS dev summit 2019 talk for more context on this work:
video: https://www.youtube.com/watch?v=JldbtDATrOo
slides: https://drive.google.com/file/d/1Ysc_3bJWmsJCETFNTRCzyvpseDpzjjf2/view

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
I've been running unit testing very similar to the test that I've added to the zfs-tests

### Future Work
Since DRR_SPILL record also (like DRR_WRITE) contains all of the data needed to recreate the damaged block - a future project could add support for healing of DRR_SPILL records.
The next logical extension for part two of this work is to provide a way for a corrupted pool to tell a backup system to generate a minimal send stream in such a way as to enable the corrupted pool to be healed with this generated send stream.
The interface could be something like the following, but maybe there are better suggestions?
```
# dumps spa err list that are part of this snapshot and the snapshot guid
zfs send -C data/fs@snap > /tmp/errlist 

# on replica system generates healing sendfile based on the errors list
zfs send -cc /tmp/errlist backup_data > /tmp/healing_sendfile

# heal our data with the minimal healing sendfile
zfs recv -c data/fs@snap < /tmp/healing_sendfile
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

alek-p commented:
There is one feature missing from this patch that I know and that is the ability to encrypt data that we have in a send file in the cases where the on-disk data is encrypted but the send file data is not.
I've yet to figure out a way to do this so the only way at this point to recv-heal encrypted datasets is with raw send streams.

megari commented:
This feature is indeed really nice to have. However, I am curious about whether it would - even theoretically - be possible to heal metadata using a corrective receive. Support for that would definitely be a killer feature.

alek-p commented:
> This feature is indeed really nice to have. However, I am curious about whether it would - even theoretically - be possible to heal metadata using a corrective receive. Support for that would definitely be a killer feature.

I agree that it would be great to be able to heal metadata but as far as I know, there isn't enough information in the send file to do that.
We are only able to heal records of type DRR_WRITE and DRR_SPILL since those are the only ones (again afaik) that contain all of the data needed to recreate damaged blocks.

alek-p commented:
I've fixed the re-encryption code so this is ready for review now.

alek-p commented:
I've rebased this to resolve the merge conflicts. I'd especially like people to look at the ZIO parts of this since I have a feeling there are better ways to perform the needed I/O. Other comments are also obviously welcomed in case anyone has some user interface improvement ideas etc. 

To help with the review you can see my OpenZFS dev summit 2019 talk for more context on this work:
- video: https://www.youtube.com/watch?v=JldbtDATrOo
- slides: https://drive.google.com/file/d/1Ysc_3bJWmsJCETFNTRCzyvpseDpzjjf2/view

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+		}
+	}
+
+	io = zio_rewrite(NULL, rwa->os->os_spa, 0, bp, abd,
+	    size, NULL, NULL, ZIO_PRIORITY_SYNC_WRITE, flags, &zb);
+
+	/* compute new bp checksum value and make sure it matches the old one */
+	zio_checksum_compute(io, BP_GET_CHECKSUM(bp), abd, size);
+	if (size != BP_GET_PSIZE(bp) ||
+	    !ZIO_CHECKSUM_EQUAL(bp_cksum, io->io_bp->blk_cksum)) {
```
this might be simplified by using `zio_checksum_error_impl()`, in which case you could verify the checksum before creation the zio, and not have to save the bp_cksum.

alek-p replied:
I'm not sure that using this rather complicate function is simpler than stashing the BP and using the macro

alek-p commented:
Thanks for the review Matt! I'll start working through these comments this weekend.

alek-p commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +130,102 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+
+/*
+ * Removes all of the recv healed errors from both on-disk error logs
+ */
+static void
+spa_remove_healed_errors(spa_t *spa, avl_tree_t *s, avl_tree_t *l, dmu_tx_t *tx)
+{
+	char name[NAME_MAX_LEN];
+	spa_error_entry_t *se;
+	void *cookie = NULL;
+
+	ASSERT(MUTEX_HELD(&spa->spa_errlog_lock));
+
+	while ((se = avl_destroy_nodes(&spa->spa_errlist_healed,
+	    &cookie)) != NULL) {
+		remove_error_from_list(spa, s, &se->se_bookmark);
+		remove_error_from_list(spa, l, &se->se_bookmark);
+		bookmark_to_name(&se->se_bookmark, name, sizeof (name));
+		kmem_free(se, sizeof (spa_error_entry_t));
+		(void) zap_remove(spa->spa_meta_objset,
+		    spa->spa_errlog_last, name, tx);
+		(void) zap_remove(spa->spa_meta_objset,
+		    spa->spa_errlog_scrub, name, tx);
+	}
+}
+
+/*
+ * Stash away healed bookmarks to remove them from the on-disk error logs
+ * later in spa_remove_healed_errors().
+ */
+void
+spa_remove_error(spa_t *spa, zbookmark_phys_t *zb)
+{
+	char name[NAME_MAX_LEN];
+
+	bookmark_to_name(zb, name, sizeof (name));
+
+	spa_add_healed_error(spa, spa->spa_errlog_last, zb);
+	spa_add_healed_error(spa, spa->spa_errlog_scrub, zb);
+}
```
I can remove the error clean up easily if we think that's cleaner.
I figured a relatively common use case for healing recv would be trying to heal using the snapshot that a scrub has IDed as corrupted. We then take this snapshot from the remote side and use it for healing. In this scenario, I figured it would make sense to then remove the errlog entry associated with the fixed blocks.

ahrens replied:
I see, that does seem like a realistic use case.

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+		ASSERT(!rwa->raw);
+		if (arc_get_compression(arc_buf) != ZIO_COMPRESS_OFF) {
+			/* Decompress the stream data */
+			err = zio_decompress_data(arc_get_compression(arc_buf),
+			    abd, buf, arc_buf_size(arc_buf), lsize);
```
fixed

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+			boolean_t no_crypt = B_FALSE;
+			dsl_pool_t *dp =
+			    rwa->os->os_dsl_dataset->ds_dir->dd_pool;
+			enc_buf = zio_data_buf_alloc(lsize);
+			abd_t *eabd = abd_get_from_buf(enc_buf, lsize);
+			abd_take_ownership_of_buf(eabd, B_FALSE);
```
that makes cleanup harder, later on, it's simple to allocate this the same way as the abd since later or I'm doing `abd = eabd;`

alek-p replied:
so it's easier to allocate `eabd` the same way as `abd`

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+			zio_crypt_decode_mac_bp(bp, mac);
+
+			dsl_pool_config_enter(dp, FTAG);
+			err = dsl_dataset_hold_flags(dp, rwa->tofs,
+			    DS_HOLD_FLAG_DECRYPT, FTAG, &ds);
```
`dsl_dataset_hold_flags(..., DS_HOLD_FLAG_DECRYPT, ...)` ensures that the keymap is created so that we can look it up later on with the `spa_keystore_lookup_key() call`

alek-p commented:
Thanks for the review Ryan, I need to expand the the testing to include spill block healing and will try to incorporate your feedback then.

alek-p commented:
I've rebased this cod, it's close to ready but it still needs more work with regards to how adb/zio is handled. There are some leaks present in the current version...

tony-zfs commented:
Hi @alek-p - in regards to the memory leaks, the receive_process_write_record (with rwa->heal set) is returning EAGAIN, but it does not add the receive_record_arg structure to the rwa->write_batch listhead. The kmem_free was also removed when it doesn't hit the EAGAIN cases for non corrective use cases.

I modified the writer thread to circumvent this situation in a local branch:

```
diff --git a/module/zfs/dmu_recv.c b/module/zfs/dmu_recv.c
index 7d29547..816632d 100644
--- a/module/zfs/dmu_recv.c
+++ b/module/zfs/dmu_recv.c
@@ -2955,9 +2955,12 @@ receive_writer_thread(void *arg)
                 * raw->write_batch), and will be used again, so we don't
                 * free it.
                 */
-               if (err != EAGAIN) {
+               if (rwa->heal) {
+                       kmem_free(rrd, sizeof (*rrd));
+               } else if (err != EAGAIN) {
                        if (rwa->err == 0)
                                rwa->err = err;
+                       kmem_free(rrd, sizeof (*rrd));
                }
        }
        kmem_free(rrd, sizeof (*rrd));
```

alek-p commented:
> Hi @alek-p - in regards to the memory leaks, the receive_process_write_record (with rwa->heal set) is returning EAGAIN, but it does not add the receive_record_arg structure to the rwa->write_batch listhead. The kmem_free was also removed when it doesn't hit the EAGAIN cases for non corrective use cases.
> 
> I modified the writer thread to circumvent this situation in a local branch:
> 
> ```
> diff --git a/module/zfs/dmu_recv.c b/module/zfs/dmu_recv.c
> index 7d29547..816632d 100644
> --- a/module/zfs/dmu_recv.c
> +++ b/module/zfs/dmu_recv.c
> @@ -2955,9 +2955,12 @@ receive_writer_thread(void *arg)
>                  * raw->write_batch), and will be used again, so we don't
>                  * free it.
>                  */
> -               if (err != EAGAIN) {
> +               if (rwa->heal) {
> +                       kmem_free(rrd, sizeof (*rrd));
> +               } else if (err != EAGAIN) {
>                         if (rwa->err == 0)
>                                 rwa->err = err;
> +                       kmem_free(rrd, sizeof (*rrd));
>                 }
>         }
>         kmem_free(rrd, sizeof (*rrd));
> ```

thanks for looking into this Tony! Afaik the only thing left now is making sure the testing is robust enough and included spill records healing.

alek-p commented:
I've hard a hard time trying to get spill block generated so I did a manual test by running the added corrective recv test with the following patch applied:
```
diff --git a/module/zfs/sa.c b/module/zfs/sa.c
index 977e729fe..8e4a8c388 100644
--- a/module/zfs/sa.c
+++ b/module/zfs/sa.c
@@ -1016,6 +1016,7 @@ sa_setup(objset_t *os, uint64_t sa_obj, sa_attr_reg_t *reg_attrs, int count,
        sa = kmem_zalloc(sizeof (sa_os_t), KM_SLEEP);
        mutex_init(&sa->sa_lock, NULL, MUTEX_NOLOCKDEP, NULL);
        sa->sa_master_obj = sa_obj;
+       sa->sa_force_spill = B_TRUE;

        os->os_sa = sa;
        mutex_enter(&sa->sa_lock);
```
I think this is ready for the next round of reviews.

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -1238,6 +1289,247 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+ * zio_rewrite the data pointed to by bp with the data from the arc buf.
+ */
+static int
+do_corrective_recv(struct receive_writer_arg *rwa, uint64_t obj,
```
I'm not sure I like that idea, that might make reading the flow harder

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+			flags |= ZIO_FLAG_RAW;
+		}
+
+		if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF) {
+			/* Recompress the data */
+			if (buf != NULL)
+				abd_free(abd);
+			size = zio_compress_data(BP_GET_COMPRESS(bp), abd, buf,
+			    lsize);
+		}
```
fixed

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -2274,7 +2645,8 @@ dmu_recv_cleanup_ds(dmu_recv_cookie_t *drc)
 		dsl_dataset_name(ds, name);
 		dsl_dataset_disown(ds, dsflags, dmu_recv_tag);
-		(void) dsl_destroy_head(name);
+		if (!drc->drc_heal)
```
we skip the destroy because we operated on an existing snapshot instead of the temp clone. I've updated the comment here: https://github.com/openzfs/zfs/pull/9372/files#diff-aa0a401dc1077010f8856e9acf77f4dcR848 to make this clearer.

alek-p commented:
Thanks for the review Paul! I think I've addressed your comments either in the PR or in the code. I've also rebased this on the latest bits from master.

PaulZ-98 commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,181 @@
+test_corrective_recv "$TESTPOOL/testfs5@snap2" $raw_backup
+
+# non raw send file healing an encrypted dataset with an unloaded key will fail
+Log_mustnot "$TESTPOOL/testfs5@snap2" $backup
```
Typo "Log_mustnot" / "log_mustnot"

freqlabs replied:
The rest of this line is not a valid command, either.

alek-p replied:
line removed. Think I accidentally pasted that in before committing

PaulZ-98 commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,181 @@
+function test_corrective_recv
+{
+	log_must zpool scrub $TESTPOOL
+	log_must zpool wait -t scrub $TESTPOOL
```
Does "zpool wait" work on FreeBSD?  From the logs it doesn't seem to be waiting. @freqlabs

freqlabs replied:
Yes, the zfs_wait tests are passing on FreeBSD.

alek-p replied:
line removed. Think I accidentally pasted that in before committing

freqlabs commented:
FWIW you may disregard the unrelated failures on the FreeBSD head bot (will be fixed by rebasing), but the zfs_receive_corrective failures do appear legitimate. I'll see if I can reproduce this locally.

freqlabs commented:
Perhaps `corrupt_blocks_at_level` is failing to corrupt the pool?
```
kern.geom.debugflags: 0 -> 16
cat: /tmp/log.29197.29197: No such file or directory
SUCCESS: dd if=/dev/urandom of=/dev/md0 bs=512 count=32 seek=8379 conv=notrunc
SUCCESS: dd if=/dev/urandom of=/dev/md0 bs=512 count=32 seek=8411 conv=notrunc
SUCCESS: dd if=/dev/urandom of=/dev/md0 bs=512 count=1 seek=3686917 conv=notrunc
SUCCESS: dd if=/dev/urandom of=/dev/md0 bs=512 count=1 seek=3686918 conv=notrunc
SUCCESS: dd if=/dev/urandom of=/dev/md0 bs=512 count=1 seek=2107998 conv=notrunc
kern.geom.debugflags: 16 -> 0
SUCCESS: zpool scrub testpool
SUCCESS: zpool wait -t scrub testpool

ERROR: eval zpool status -v testpool |      grep "Permanent errors have been detected" exited 1
```

`cat: /tmp/log.29197.29197: No such file or directory` looks very odd.

freqlabs commented:
> `cat: /tmp/log.29197.29197: No such file or directory` looks very odd.

This is from logapi.shlib, some command must be so malformed... ah so there are a few bugs in `corrupt_blocks_at_level` and `list_file_blocks`, I'll have a go at fixing that up!

For the curious, I started with this change, which unmasked the errors:
```diff
--- a/tests/zfs-tests/include/blkdev.shlib
+++ b/tests/zfs-tests/include/blkdev.shlib
@@ -602,7 +602,7 @@ function corrupt_blocks_at_level # input_file corrupt_level
                sysctl kern.geom.debugflags=16
        fi

-       log_must list_file_blocks $input_file | \
+       list_file_blocks $input_file | \
            while read level path offset length; do
                if [[ $level = $corrupt_level ]]; then
                        log_must dd if=/dev/urandom of=$path bs=512 \
```
```
kern.geom.debugflags: 0 -> 16
stat: illegal option -- c
usage: stat [-FLnq] [-f format | -l | -r | -s | -x] [-t timefmt] [file|handle ...]
```

freqlabs commented:
https://github.com/openzfs/zfs/pull/10572

freqlabs commented:
Now it fails here:
```
SUCCESS: zfs unmount testpool/testfs5
SUCCESS: zfs unload-key testpool/testfs5
SUCCESS: zpool scrub testpool
SUCCESS: zpool wait -t scrub testpool
errors: Permanent errors have been detected in the following files:
SUCCESS: eval zpool status -v testpool |            grep "Permanent errors have been detected"
SUCCESS: zinject -a
SUCCESS: eval zfs recv -c testpool/testfs5@snap2 < /var/tmp/raw_backup
SUCCESS: zpool sync testpool
SUCCESS: eval zpool status -v testpool |            grep "Permanent errors have been detected" exited 1
/home/ryan/ZoF/tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh[162]: log_mustnot[152]: log_neg[57]: log_neg_expect[206]: testpool/testfs5@snap2: not found [No such file or directory]
ERROR: testpool/testfs5@snap2 /var/tmp/backup unexpectedly exited 127 (File not found)
```

alek-p commented:
thanks for looking into the FreeBSD failures. I've fixed the problem line and rebased but this still won't work until #10572 is merged at which point I'll rebase again

alek-p commented:
Seems like this is pretty much ready except for the FreeBSD test that's causing a panic. I've not been able to figure that out yet. It's the `redacted_send/redacted_panic.ksh`

PaulZ-98 commented:
Hey Alek thanks for working on this!  I think to pass the FreeBSD test you might need to back port all the commits of https://github.com/openzfs/zfs/pull/11297 to your branch, because it looks like the test that's panic-ing is probably the one added to check for that bug fix.

alek-p commented:
> Hey Alek thanks for working on this! I think to pass the FreeBSD test you might need to back port all the commits of #11297 to your branch, because it looks like the test that's panic-ing is probably the one added to check for that bug fix.

I thought that at first but I rebased 4 days ago so I've picked that up already.

freqlabs commented:
Are you able to reproduce the test failure in a FreeBSD VM?

alek-p commented:
> Are you able to reproduce the test failure in a FreeBSD VM?

Yeah, my VM [panics](https://ibb.co/XLYXpx5) every time if I run `redacted_send/redacted_panic.ksh` with this patch. Seems like it's a timing thing since the FreeBSD 12 tests passed.

freqlabs commented:
<img width="1444" alt="Screen Shot 2021-04-27 at 3 39 28 PM" src="https://user-images.githubusercontent.com/5410316/116302081-c3e8ff00-a76e-11eb-98ca-6adbf351b9e4.png">
I'm guessing that second number is probably unreasonable for this test?

PaulZ-98 commented:
I was looking to see if bytes_read (or what it's set from) was uninitialized somewhere but didn't find anything obvious.

alek-p commented:
Thanks Paul. I think the wrong value is coming from disk when we're reading DS_FIELD_RESUME_BYTES.

PaulZ-98 commented:
I wonder if this line should be && not || ?  i.e. are they always both maintained at 0 for "no resume"?

{code}
lzc_send_resume_redacted(..)
...
...
	if (resumeobj != 0 || resumeoff != 0) {
		fnvlist_add_uint64(args, "resume_object", resumeobj);
		fnvlist_add_uint64(args, "resume_offset", resumeoff);
	}
{\code}

same thing in lzc_send_space_resume_redacted

alek-p commented on module/zfs/dmu_recv.c:
```diff
@@ -2881,7 +3191,7 @@ dmu_recv_stream(dmu_recv_cookie_t *drc, offset_t *voffp)
 
 	if (dsl_dataset_is_zapified(drc->drc_ds)) {
-		uint64_t bytes;
+		uint64_t bytes = 0;
```
this was the fix for the FreeBSD `redacted_panic.ksh` test panic. It turns out the panic wasn't related to my patch, but the patch did happen to trigger the issue more reliably.

Big thanks to @PaulZ-98 for tracking down the cause of the panic. All tests are now passing.

clhedrick commented:
If it's not there, I'd like to request an option to do the comparison but simply report when there's a mismatch. Many of us use zfs send streams for backups. There's no real choice for file systems with huge numbers of small files, as tar, etc, are too slow. With this option, it could be used to verify a backup, which is a major missing function.

alek-p commented:
> If it's not there, I'd like to request an option to do the comparison but simply report when there's a mismatch. Many of us use zfs send streams for backups. There's no real choice for file systems with huge numbers of small files, as tar, etc, are too slow. With this option, it could be used to verify a backup, which is a major missing function. This is a feature you'd probably want anyway, since you might want to assess what it's going to do before doing it.

By "do the comparison but simply report when there's a mismatch" I assume you mean reading data from disk and reporting when there is a checksum error with these reads? Sounds like you want a (per dataset?) scrub.
If you just want to verify the integrity of a backup, a scrub is better since it also checks the integrity of the metadata and this patch ignores metadata since it's not easy to heal it using a send stream.

clhedrick commented:
No. The use case is verifying a tape backup, which is a zfs send stream. I'd like to compare it with the source and verify that there are no differences. You'd do a correction receive to the original file system with the option to report differences rather than correct them.

clhedrick commented:
The other half of tape backups (which I'm not asking you to fix here) is for receive to skip things it can't read.  At the moment a send stream backup is less robust than a tar backup, because if the tape is partially damaged you lose everything after the error.

alek-p commented:
> No. The use case is verifying a tape backup, which is a zfs send stream. I'd like to compare it with the source and verify that there are no differences. You'd do a correction receive to the original file system with the option to report differences rather than correct them.

So you want to verify the integrity of the send stream? [zstreamdump](https://openzfs.github.io/openzfs-docs/man/8/zstreamdump.8.html) does that.

Since corrective receive ignores all records except DRR_WRITE records it would not be verifying metadata records so this is not the right place to do stream validation. You can use zstreamdump to validate checksums for all records without involving the source.

fredcooke commented:
I could have done with this last night. I was going to write up an issue describing my experience with a slave pool that has never seen any direct change except two mount points set differently to the up stream pool and how it had 5 "permanent" errors that persisted after a successful ~400GB send/receive incremental but cleaned up after the double scrub and how I thought a double scrub was too heavy for 5 tiny issues on a file system that I couldn't fault for content on its read only mount. And how the oracle docs have 3 extra arguments for `zpool clear` that openzfs clear doesn't have that sound useful (roll back to an earlier state to recover) which would have suited me as I just wanted minimal disk activity path to clean `status -v` and I click on the `6zb6AxL26d8` yt vid and look in the meeting minutes doc and here's my PR. I will attempt to look over your changes despite never having read any of the OpenZFS source as I think it's a big deal and a huge step in the right direction for my use case at least.

fredcooke commented on man/man8/zfs-receive.8:
```diff
@@ -379,6 +384,22 @@ restrictions (e.g. set-once) apply equally to
+by using the provided stream as the source of healthy data. This method of
+healing can only heal data blocks present in the stream. Metadata can not
+be healed by corrective receive. Running a scrub is recommended post
+healing to ensure all corruption had been healed. It's important to consider
+why corruption has happened in the first place since if you have slowly failing
+hardware periodically healing the data is not going to save you from data loss
+later on when the hardware fails completeley.
```
Good advice, thank you.

fredcooke commented on man/man8/zfs-receive.8:
```diff
@@ -379,6 +384,22 @@ restrictions (e.g. set-once) apply equally to
+.Nm snapshot,
+by using the provided stream as the source of healthy data. This method of
+healing can only heal data blocks present in the stream. Metadata can not
+be healed by corrective receive. Running a scrub is recommended post
```
That's disappointing. Why can metadata not be healed in this way? Is there any other existing way it can be healed or could that facility be added to this PR?

fredcooke replied:
`The other thing to point out is that we can only correct data. Specifically, we are only able to heal records of type DRR_WRITE and DRR_SPILL since those are the only ones that contain all of the data needed to recreate the damaged block.` from the body of the PR. I guess my real question is "how can I correct these "small errors" without a full disk read worth of scrub times two? But it doesn't belong here so I'll make a new issue to describe my situation and link to other similar things I found last night while unable to sleep.

fredcooke commented on lib/libzfs/libzfs_sendrecv.c:
```diff
@@ -4393,9 +4416,34 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
+		    get_snap_guid(hdl, name, strchr(destsnap, '@') + 1);
+		if (guid == 0) {
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "corrective recv must specify an existing snapshot"
+			    " to heal"));
```
How does one know which snapshot(s) need healing from a set of errors like this: 

```errors: Permanent errors have been detected in the following files:

        <metadata>:<0x0>
        <metadata>:<0x1>
        <metadata>:<0x3f3>
        /rust/private/
        whirlpool/data:<0x0>
```

And would it be worthwhile adding some kind of instructions like that to the manpage change further down or does it belong elsewhere?

fredcooke replied:
`There are likely several snapshots referring to the same potentially corrupted data so there may be many snapshots with the above condition holding that are able to heal a single block.` this comment from the PR body partially answers my question. But in my case the only tangible thing being `/rust/private` had nothing in it prior to the most recent send/receive except other datasets and post that send/receive had one folder which seemed fine (did not do a diff -rq to check mind you). So I'm guessing that in my case given I have 3 metadata errors and one empty dataset error and one empty mount point error, there were no data blocks and this PR couldn't have helped me anyway, is that right?

alek-p replied:
yup, this PR would not have healed your metadata errors

fredcooke commented:
Final two thought for now:

1. Is there a companion PR to the OpenZFS github hosted documentation to go with this change? I feel like that should exist, too?
2. Can you please put a link to this PR in the tail end of the commit comment as a reference to the superior description herein that's not in your commit comment itself, next time you rebase and force push? I feel like that's a valuable cross ref to have. 

Cheers!

fredcooke commented:
@ahrens (or anyone else such as @mmaybee) can you please reference the old incomplete Delphix intern PR about producing/displaying info about the permanent errors and what they're a part of/what they affect/what they are as mentioned at 5:56 in `6zb6AxL26d8` - I'm interested in being mentored on that but I looked and I couldn't find it. I did find this semi related awesome PR #8995 about partial focussed scrubs, though, that alone would have solved my issue last night esp if `status` had told me to run it.

alek-p commented:
> Final two thought for now:
> 
> 1. Is there a companion PR to the OpenZFS github hosted documentation to go with this change? I feel like that should exist, too?
> 2. Can you please put a link to this PR in the tail end of the commit comment as a reference to the superior description herein that's not in your commit comment itself, next time you rebase and force push? I feel like that's a valuable cross ref to have.
> 
> Cheers!

1) Not sure what documentation changes you're looking for but once this is merged the man page for [recv](https://openzfs.github.io/openzfs-docs/man/8/zfs-receive.8.html) will be auto-updated with the man page changes proposed in this patch.
2) Once this is merged the commit will have a link to this PR at the bottom (see any other merged commits for an [example](https://github.com/openzfs/zfs/commit/fce29d6aa4b4067cd5f0d6cd7481cce469dca3e8)).

fredcooke commented:
1. I see that the man pages are part of that site I was talking about, hadn't realised that until now. So all good on that front. Might be worth fleshing out the man pages with more "guide like" stuff across the board and treating it as a user manual rather than a sparse "app spec" like most man pages are. 
2. Fair enough, I always strictly follow a fast forward workflow even on GH the last service that does not allow it using the UI. Especially for branches like this with one commit the signal to noise ratio of merge commits is 50% and if most of the PRs are 1 commit that makes for an intolerably low density history. But on busy fast moving open sauce projects it remains the only thing that makes sense.

PaulZ-98 commented on module/zfs/dmu_recv.c:
```diff
@@ -1236,6 +1289,185 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+		cb_data->spa = rwa->os->os_spa;
+		cb_data->size = drrw->drr_logical_size;
+		cb_data->zb = zb;
+		/* Test if healing worked by re-reading the bp */
```
If this re-read generates ECKSUM because the healing didn't work, it just stays as a checksum error?  Maybe the zfstest can cover this?

alek-p replied:
It does stay a checksum error yeah. I'm not sure how I could produce a healing failure for zfs-test, any ideas on that?

alek-p replied:
The way it's written now this function does return an error if the zio_read() fails and zfs_recv_best_effort_corrective is not set to zero (which is the default). For testing, I couldn't come up with a way to selectively corrupt a healed block but I did test this path by just forcing the functions to return ECKSUM as you suggested and the results were reasonable:
The corrective recv returned `cannot receive new filesystem stream: corrective receive was not able to reconstruct the data needed for healing.` msg and I was able to unload the module ok (without leak etc)

PaulZ-98 replied:
I'm only seeing "err" set from the zio_rewrite.  The zio_read is with zio_nowait, so I don't see it setting/returning err.  It's good that corrective_read_done preserves the spa error if the zio_read fails though.

alek-p replied:
oh i see what you mean now. I wasn't sure if it was better to use async re-read there but since we just wrote that block it probably does make sense to wait for the re-read and return it's results. I've updated the code to do that.
The only downside I can think of is if there is a lot of corruption the healing recv will run slower if we're not using async re-reads.

PaulZ-98 commented on module/zfs/dmu_recv.c:
```diff
@@ -2043,6 +2275,60 @@ receive_process_write_record(struct receive_writer_arg *rwa,
+		 */
+		if (err != ECKSUM)
+			return (err);
+		err = dmu_buf_hold_noread(rwa->os, drrw->drr_object,
```
Could the dmu_read and dmu_buf_hold_noread just be replaced by a single dmu_buf_hold (since it tries the dmu_buf_hold_noread) and then test for err != ECKSUM, then proceed on using the dbp you filled in?

alek-p replied:
Can't use `dmu_buf_hold()` since it zeros out `dbp` when we get an error (`ECKSUM` for example). However due looking at this some more I've update the code to do the `dmu_read()` after the `dmu_hold_noread()` which is what `dmu_buf_hold()` does.

PaulZ-98 commented on module/zfs/dmu_recv.c:
```diff
@@ -2043,6 +2275,60 @@ receive_process_write_record(struct receive_writer_arg *rwa,
+			return (err);
+
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(rwa->os, drrw->drr_object, drrw->drr_offset,
```
This can be dbuf_read which is a little cleaner.
```
diff --git a/module/zfs/dmu_recv.c b/module/zfs/dmu_recv.c
index 5d7c9ad88..595e01df2 100644
--- a/module/zfs/dmu_recv.c
+++ b/module/zfs/dmu_recv.c
@@ -2279,7 +2279,6 @@ receive_process_write_record(struct receive_writer_arg *rwa,
                blkptr_t *bp;
                dmu_buf_t *dbp;
                dnode_t *dn;
-               uint8_t buf[1];
                int flags = DMU_READ_PREFETCH;
 
                if (rwa->raw)
@@ -2298,10 +2297,10 @@ receive_process_write_record(struct receive_writer_arg *rwa,
                        return (err);
 
                /* Try to read the object to see if it needs healing */
-               err = dmu_read(rwa->os, drrw->drr_object, drrw->drr_offset,
-                   sizeof (buf), buf, flags);
+               err = dbuf_read((dmu_buf_impl_t *)dbp, NULL, DB_RF_CANFAIL);
+
                /*
-                * We only try to heal when dmu_read() returns a ECKSUMs.
+                * We only try to heal when dbuf_read() returns ECKSUM.
                 * Other errors (even EIO) get returned to caller.
                 * EIO indicates that the device is not present/accessible,
                 * so writing to it will likely fail.
```

alek-p commented:
I've now added the new corrective recv ioctl to py-libzfs

alek-p commented:
added a test that makes sure that incremental recv still works after a corrective recv: https://github.com/openzfs/zfs/pull/9372/files#diff-84216657d22949aaaa50ebed2ffcf6378ff0f97886165111c85e3e981c7bb550R168-R172

alek-p commented:
The last thing left to sort out if why during a zio_rewrite() os_meta_dnode supposed is also updated. This is causing changes to the snapshot so that we can't do incremental recv on top of it without doing a rollback first.

alek-p commented:
Thanks for @PaulZ-98 for figuring out the incremental recv "issue". 
The problem was the way the test is set up it does an md5 sum on a file and with atime=on this was causing modifications to the source dataset making incremental recv not possible. I changed the test to create the testing filesystem with atime=off to avoid this behavior. Alternatively, we can get rid of the md5 checksum checking part of the test instead.
This is ready for review once again.

PaulZ-98 reviewed (approved):
Looks good.

alek-p commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,192 @@
+	log_fail "Unexpected compression $compr in recved dataset"
+corrupt_blocks_at_level "/$TESTPOOL/testfs3/$TESTFILE0" 0
+# test healing recv when compression on-disk is on but source was uncompressed
+test_corrective_recv "$TESTPOOL/testfs3@snap1" $backup
```
this needs to be `$unc_backup` to match the comment and intent of the test. I'll fix it right now

PaulZ-98 commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,193 @@
+
+#
+# Copyright (c) 2019 Datto, Inc. All rights reserved.
+# Copyrgith (c) 2020 Axcient. All rights reserved.
```
Typo Copyright

alek-p replied:
thanks, fixed

nabijaczleweli commented on man/man8/zfs-receive.8:
```diff
@@ -57,6 +57,11 @@
+.Cm receive
+.Fl c
+.Op Fl vn
+.Ar filesystem Ns | Ns Ar snapshot
```
You dropped the `.` before `.Sh`.

nabijaczleweli commented on man/man8/zfs-receive.8:
```diff
@@ -393,6 +398,23 @@ restrictions (e.g. set-once) apply equally to
+Metadata can not be healed by corrective receive.
+Running a scrub is recommended post healing to ensure all corruption had
+been healed.
+It's important to consider why corruption has happened in the first place
```
This wants a `.Pp` before the "It's".

nabijaczleweli commented on module/zfs/dmu_recv.c:
```diff
@@ -67,6 +70,7 @@
 int zfs_recv_queue_length = SPA_MAXBLOCKSIZE;
 int zfs_recv_queue_ff = 20;
 int zfs_recv_write_batch_size = 1024 * 1024;
+int zfs_recv_best_effort_corrective = 0;
```
`static`

alek-p replied:
think there is a reason none of the tunables are static so that's why the new one isn't static either

nabijaczleweli replied:
That reason is purely because #12899 wasn't merged yet. I don't necessarily feel like re-running that on new code when it is.

alek-p replied:
Looks like I would need to wait for #12899 to land before I can make the tunable static. If your PR lands before this patch I'll rebase and add the static keyword.

nabijaczleweli commented on man/man4/zfs.4:
```diff
@@ -1848,6 +1848,16 @@ This setting will not reduce the write size below a single block.
 .Sy 32MB .
 .
+.It Sy zfs_recv_best_effort_corrective Ns = Ns Sy 0 Ns Pq int
+When this variable is set to non-zero a corrective receive:
+  1) Does not enforce the restriction of source & destination snapshot GUIDs
+  matching.
+  2) If there is an error during healing, the healing receive is not
+  terminated instead it moves on to the next record.
+.
+Default value is
+.Sy 0 .
```
(a) make this a list, (b) drop the default value sentence, (c) drop the Ns before Pq

nabijaczleweli commented on man/man4/zfs.4:
```diff
@@ -1848,6 +1848,17 @@ This setting will not reduce the write size below a single block.
 .Sy 32MB .
 .
+.It Sy zfs_recv_best_effort_corrective Ns = Ns Sy 0 Pq int
+.Bl -tag -compact
```
-bullet or -enum (rest of zfs.4 uses -enum, so copy that whole line)

alek-p replied:
fixed. Thanks for the reviews

alek-p commented:
https://github.com/openzfs/zfs/commit/d0d291d36747a1e0ce3d5ffb06f38cd700ee634e - rebased to master

jumbi77 commented:
> [d0d291d](https://github.com/openzfs/zfs/commit/d0d291d36747a1e0ce3d5ffb06f38cd700ee634e) - rebased to master

Just want to thanks for your continuously rebasing (since late 2019 and not losing interest in upstreaming this enhancement). This is a nice feature which is definitely worth getting upstreamed! Much thanks!

@behlendorf @ahrens @freqlabs can I may/politely ping you to review this?

Thanks to all participants!

alek-p commented:
I've rebased this after the new errlog format patch was merged (https://github.com/openzfs/zfs/pull/12812).
The corrective recv test is failing if ran on a pool with errlog feature enabled because corruption is not reported in `zpool status -v` output (even though the vdev shows errors) when we scrub a corrupted pool after unloading the encryption key. I'm thinking this is an errlog feature bug but I haven't had time to look into it. 
Until I can confirm and write up bug report I've changed the test to create the pool with errlog=disabled.

pcd1193182 commented on module/zfs/dmu_recv.c:
```diff
@@ -1243,6 +1300,182 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+			flags |= ZIO_FLAG_RAW;
+		}
+
+		if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF) {
```
Probably a silly question, but doesn't this result in compressing the already encrypted data (which is going to not be compressible)? Don't we want the other order?

alek-p replied:
Good point, yeah the order was wrong I've fixed that but it got me wondering why testing didn't catch that. And then I realized that I'm generating incompressible data. Fixing that revealed another bug. Thanks for taking a look

behlendorf commented:
Would you mind rebasing this again to resolve the build system conflicts.  Then I think we should be close, I'd just be good to understand what's going on with the errlog feature and the test case.

alek-p commented:
> Would you mind rebasing this again to resolve the build system conflicts. Then I think we should be close, I'd just be good to understand what's going on with the errlog feature and the test case.

After @pcd1193182's review, I realized that the testing wasn't generating compressible data. Fixing that revealed a hang so I'll need to spend some time digging into it before I'm ready to rebase this.

alek-p commented:
> Would you mind rebasing this again to resolve the build system conflicts. Then I think we should be close, I'd just be good to understand what's going on with the errlog feature and the test case.

I've rebased this on master and think it's ready for review again. I did check the tests with the errlog feature enabled and it's still not showing the corruption even though there are checksum errors being reported.

I've restructured the code a little and fixed a couple of decompression bugs. Thanks to @PaulZ-98 (once again) for helping me track down the hang!

alek-p commented:
I've added one more test. It makes sure we can use send files that were generated with `zfs send --large-block` for corrective recv.

alek-p commented:
> I've added one more test. It makes sure we can use send files that were generated with `zfs send --large-block` for corrective recv.

had to remove this due to https://github.com/openzfs/zfs/issues/13699 will add again one that is resolved

t-oster commented:
@alek-p congratulations 🎉 and thanks for your endurance

pepsinio commented:
Would this fix make it to release soon?

GregorKopka commented:
> Would this fix make it to release soon?

@behlendorf ?

behlendorf commented:
This feature will make it in to the OpenZFS 2.2 release.

pepsinio commented:
> 

This is a huge improvement. The last missing piece for my use case with offsite backup, no RaidZ and having low throughput link which makes it difficult to recreate datasets with ease in case of errors. Quite certain i am not the only one on this boat.

FlorianHeigl commented:
@pepsinio you're not the only one on that boat. I follow this topic since before the PR was opened. no WAN use case but the generally understanding that this is a major resilience feature that stabilizes many related bits and bytes.
